File: sirius-core/src/main/java/org/starcoin/sirius/core/IOU.java
package org.starcoin.sirius.core;

import org.starcoin.core.ProtobufCodec;
import org.starcoin.proto.Starcoin.ProtoIOU;

public class IOU implements ProtobufCodec<ProtoIOU> {

    private OffchainTransaction transaction;

    private Update update;

    public IOU() {
    }

    public IOU(OffchainTransaction transaction, Update update) {
        this.transaction = transaction;
        this.update = update;
    }

    public IOU(ProtoIOU protoIOU) {
        this.unmarshalProto(protoIOU);
    }

    public OffchainTransaction getTransaction() {
        return transaction;
    }

    public void setTransaction(OffchainTransaction transaction) {
        this.transaction = transaction;
    }

    public Update getUpdate() {
        return update;
    }

    public void setUpdate(Update update) {
        this.update = update;
    }

    @Override
    public ProtoIOU marshalProto() {
        return ProtoIOU.newBuilder().setTransaction(this.transaction.marshalProto())
                .setUpdate(this.update.marshalProto()).build();
    }

    @Override
    public void unmarshalProto(ProtoIOU proto) {
        this.transaction = new OffchainTransaction(proto.getTransaction());
        this.update = new Update(proto.getUpdate());
    }
}

File: sirius-core/src/main/kotlin/org/starcoin/sirius/core/IOU.kt
package org.starcoin.sirius.core

import kotlinx.serialization.SerialId
import kotlinx.serialization.Serializable
import org.starcoin.proto.Starcoin
import org.starcoin.sirius.serialization.ProtobufSchema

@Serializable
@ProtobufSchema(Starcoin.IOU::class)
data class IOU(@SerialId(1) val transaction: OffchainTransaction, @SerialId(2) val update: Update) :
    SiriusObject() {

    companion object : SiriusObjectCompanion<IOU, Starcoin.IOU>(IOU::class) {
        override fun mock(): IOU {
            return IOU(OffchainTransaction.mock(), Update.mock())
        }

        override fun parseFromProtoMessage(proto: Starcoin.IOU): IOU {
            return IOU(
                OffchainTransaction.parseFromProtoMessage(proto.transaction),
                Update.parseFromProtoMessage(proto.update)
            )
        }

        override fun toProtoMessage(obj: IOU): Starcoin.IOU {
            return Starcoin.IOU.newBuilder().setTransaction(OffchainTransaction.toProtoMessage(obj.transaction))
                .setUpdate(Update.toProtoMessage(obj.update)).build()
        }
    }
}
